src/HealthGPS/kevin_hall_model.cpp

```diff
@@ -114,6 +139,89 @@ void KevinHallModel::update_risk_factors(RuntimeContext &context) {
     }
 }
 
+void KevinHallModel::initialise_sector(RuntimeContext &context, Person &person) const {
+
+    // Get rural prevalence for age group and sex.
+    double prevalence;
+    if (person.age < 18) {
+        prevalence = rural_prevalence_.at("Under18"_id).at(person.gender);
+    } else {
+        prevalence = rural_prevalence_.at("Over18"_id).at(person.gender);
+    }
+
+    // Sample the person's sector.
+    double rand = context.random().next_double();
+    auto sector = rand < prevalence ? core::Sector::rural : core::Sector::urban;
+    person.sector = sector;
+}
+
+void KevinHallModel::update_sector(RuntimeContext &context, Person &person) const {
+
+    // Only update rural sector 18 year olds.
+    if ((person.age != 18) || (person.sector != core::Sector::rural)) {
+        return;
+    }
+
+    // Get rural prevalence for age group and sex.
+    double prevalence_under18 = rural_prevalence_.at("Under18"_id).at(person.gender);
+    double prevalence_over18 = rural_prevalence_.at("Over18"_id).at(person.gender);
+
+    // Compute random rural to urban transition.
+    double rand = context.random().next_double();
+    double p_rural_to_urban = 1.0 - prevalence_over18 / prevalence_under18;
+    if (rand < p_rural_to_urban) {
+        person.sector = core::Sector::urban;
+    }
+}
+
+void KevinHallModel::initialise_income(RuntimeContext &context, Person &person) const {
+
+    // Compute logits for each income category.
+    auto logits = std::vector<double>(income_models_.size());
+    for (size_t i = 0; i < income_models_.size(); i++) {
+        logits[i] = income_models_[i].intercept;
+        for (const auto &[factor_name, coefficient] : income_models_[i].coefficients) {
+            logits[i] += coefficient * person.get_risk_factor_value(factor_name);
+        }
+    }
```

Comment:
Alternative implementation using ranges (you'll also need `#include <ranges>`):

```suggestion
    auto logits = std::vector<double>{};
    logits.reserve(income_models_.size());
    std::ranges::transform(income_models_, std::back_inserter(logits),
                           [&person](const auto &model) {
                               double logit = model.intercept;
                               for (const auto &[factor_name, coefficient] : model.coefficients) {
                                   logit += coefficient * person.get_risk_factor_value(factor_name);
                               }
                               return logit;
                           });
```